node/src/components/linear_chain_sync.rs

```diff
@@ -449,14 +506,34 @@ where
                         }
                     }
                     BlockByHeightResult::FromStorage(block) => {
-                        // We shouldn't get invalid data from the storage.
-                        // If we do, it's a bug.
-                        assert_eq!(block.height(), block_height, "Block height mismatch.");
-                        trace!(%block_height, "Linear block found in the local storage.");
-                        // When syncing descendants of a trusted hash, we might have some of them in
-                        // our local storage. If that's the case, just
-                        // continue.
-                        self.block_downloaded(rng, effect_builder, &block)
+                        match self.state {
+                            State::SyncingTrustedHash {
+                                highest_block_header: Some(ref hdr),
+                                ref validator_weights,
+                                ..
+                            } => {
+                                info!(%block_height, "Read a block from storage while syncing up\
+                                    to trusted hash; Starting at the highest known block\
+                                    instead.");
+                                let peer = self.peers.random_unsafe();
+                                // Kick off syncing trusted hash descendants.
+                                let height = hdr.height();
+                                let hash = hdr.hash();
+                                let weights = validator_weights.clone();
+                                self.state = State::sync_descendants(hash, *block, weights);
```

Comment:
When we're downloading blocks by height we're already in the `SyncingDescenandts` mode. I don't understand why we need this at all  :grimacing: :see_no_evil:

Reply:
Huh, good point. I probably intended to put this code in the place where we handle the result of getting a block by hash. But if it works as it is, this part might not be needed at all :sweat_smile:

Reply:
Can you please check if it still works if you remove it? :)

Reply:
The reason this piece is not needed is b/c of the change in `block_downloaded` method – when we're in `SyncingTrustedHash` state we're detecting whether the new block is a child of our highest known block. If so, then we start downloading its deploys. It's actually exactly what I wanted to do :+1: